eclair-core/src/main/scala/fr/acinq/eclair/transactions/Transactions.scala

```diff
@@ -268,25 +334,32 @@ object Transactions {
-    var outputsAlreadyUsed = Set.empty[Int] // this is needed to handle cases where we have several identical htlcs
-    val htlcTimeoutTxs = trimOfferedHtlcs(localDustLimit, spec).map { htlc =>
-      val htlcTx = makeHtlcTimeoutTx(commitTx, outputsAlreadyUsed, localDustLimit, localRevocationPubkey, toLocalDelay, localDelayedPaymentPubkey, localHtlcPubkey, remoteHtlcPubkey, spec.feeratePerKw, htlc.add)
-      outputsAlreadyUsed = outputsAlreadyUsed + htlcTx.input.outPoint.index.toInt
-      htlcTx
+  def makeHtlcTxs(commitTx: Transaction,
+                  localDustLimit: Satoshi,
+                  localRevocationPubkey: PublicKey,
+                  toLocalDelay: CltvExpiryDelta,
+                  localDelayedPaymentPubkey: PublicKey,
+                  feeratePerKw: Long,
+                  outputs: Seq[CommitmentOutputLink]): (Seq[HtlcTimeoutTx], Seq[HtlcSuccessTx]) = {
+
+    val htlcTimeoutTxs = outputs.zipWithIndex.collect {
+      case (co@CommitmentOutputLink(_, _, DirectedHtlc(OUT, _)), outputIndex) =>
+        makeHtlcTimeoutTx(commitTx, co, outputIndex, localDustLimit, localRevocationPubkey, toLocalDelay, localDelayedPaymentPubkey, feeratePerKw)
     }
-    val htlcSuccessTxs = trimReceivedHtlcs(localDustLimit, spec).map { htlc =>
-      val htlcTx = makeHtlcSuccessTx(commitTx, outputsAlreadyUsed, localDustLimit, localRevocationPubkey, toLocalDelay, localDelayedPaymentPubkey, localHtlcPubkey, remoteHtlcPubkey, spec.feeratePerKw, htlc.add)
-      outputsAlreadyUsed = outputsAlreadyUsed + htlcTx.input.outPoint.index.toInt
-      htlcTx
+    val htlcSuccessTxs = outputs.zipWithIndex.collect {
+      case (co@CommitmentOutputLink(_, _, DirectedHtlc(IN, _)), outputIndex)  =>
+        makeHtlcSuccessTx(commitTx, co, outputIndex, localDustLimit, localRevocationPubkey, toLocalDelay, localDelayedPaymentPubkey, feeratePerKw)
     }
     (htlcTimeoutTxs, htlcSuccessTxs)
   }
 
-  def makeClaimHtlcSuccessTx(commitTx: Transaction, outputsAlreadyUsed: Set[Int], localDustLimit: Satoshi, localHtlcPubkey: PublicKey, remoteHtlcPubkey: PublicKey, remoteRevocationPubkey: PublicKey, localFinalScriptPubKey: ByteVector, htlc: UpdateAddHtlc, feeratePerKw: Long): ClaimHtlcSuccessTx = {
+  def makeClaimHtlcSuccessTx(commitTx: Transaction, outputs: Seq[CommitmentOutputLink], localDustLimit: Satoshi, localHtlcPubkey: PublicKey, remoteHtlcPubkey: PublicKey, remoteRevocationPubkey: PublicKey, localFinalScriptPubKey: ByteVector, htlc: UpdateAddHtlc, feeratePerKw: Long): ClaimHtlcSuccessTx = {
     val redeemScript = htlcOffered(remoteHtlcPubkey, localHtlcPubkey, remoteRevocationPubkey, ripemd160(htlc.paymentHash.bytes))
-    val pubkeyScript = write(pay2wsh(redeemScript))
-    val outputIndex = findPubKeyScriptIndex(commitTx, pubkeyScript, outputsAlreadyUsed, amount_opt = Some(htlc.amountMsat.truncateToSatoshi))
+    val Some(outputIndex) = outputs.zipWithIndex.find {
```

Comment:
There is a small regression here. If for some reason the htlc matching fails, we previously threw an `OutputNotFound` (in `findPubKeyScriptIndex`). Now it's going to be a pattern matching failure on the `Some`, and it's a bit less explicit.

It seems to me that an exception here would be really bad, as it would prevent us from reacting to a remote commitment spent on-chain. Why not make this function return an `Option[ClaimHtlcSuccessTx]` and log a big error message where it's called in `Helpers.scala` if it returns `None`? This way we would miss claiming the htlc success, but at least we'd be able to claim our main output. Maybe it's a bad idea and we'd prefer eclair to crash in that case to force the user to react, let me know your opinion.

Same for the `claimHtlcTimeoutTx` below.

Reply:
Yes throwing an exception would be pretty bad if it interrupts the execution flow, in fact we always "nest" the transaction generation within a `try/catch` in the [`generateTx` function](https://github.com/ACINQ/eclair/blob/33e9ac56c665573ac6f2c25aeadcd75a152fc6a3/eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala#L618), this is used also when calling `makeClaimHtlcSuccessTx` to generate the HTLC-success tx.

Reply:
Oh yeah since this is in a Try it's ok then.
We can make this return an `Option` in a later PR to keep this focused.